request.go

```diff
@@ -3,376 +3,145 @@ package cmds
-func (r *request) SetOptions(opts cmdkit.OptMap) error {
-	r.options = opts
-	return r.ConvertOptions()
-}
-
-func (r *request) StringArguments() []string {
-	return r.arguments
-}
-
-// Arguments returns the arguments slice
-func (r *request) Arguments() []string {
-	if r.haveVarArgsFromStdin() {
-		err := r.VarArgs(func(s string) error {
-			r.arguments = append(r.arguments, s)
-			return nil
-		})
-		if err != nil && err != io.EOF {
-			log.Error(err)
-		}
+	fi, err := req.Files.NextFile()
+	if err != nil {
+		return nil, err
 	}
 
-	return r.arguments
-}
-
-func (r *request) SetArguments(args []string) {
-	r.arguments = args
-}
-
-func (r *request) Files() files.File {
-	return r.files
+	return bufio.NewScanner(fi), nil
 }
 
-func (r *request) SetFiles(f files.File) {
-	r.files = f
+func IsAllArgsAlreadyCovered(err error) bool {
+	return err != nil && err.Error() == "all arguments covered by positional arguments"
 }
 
-func (r *request) Context() context.Context {
-	return r.rctx
+func IsMoreArgumentsExpected(err error) bool {
+	return err != nil && err.Error() == "expected more arguments from stdin"
 }
 
-func (r *request) haveVarArgsFromStdin() bool {
-	// we expect varargs if we have a string argument that supports stdin
-	// and not arguments to satisfy it
-	if len(r.cmd.Arguments) == 0 {
-		return false
-	}
-
-	last := r.cmd.Arguments[len(r.cmd.Arguments)-1]
-	return last.SupportsStdin && last.Type == cmdkit.ArgString && (last.Required || last.Variadic) &&
-		len(r.arguments) < len(r.cmd.Arguments)
-}
-
-// VarArgs can be used when you want string arguments as input
-// and also want to be able to handle them in a streaming fashion
-func (r *request) VarArgs(f func(string) error) error {
-	if len(r.arguments) >= len(r.cmd.Arguments) {
-		for _, arg := range r.arguments[len(r.cmd.Arguments)-1:] {
-			err := f(arg)
-			if err != nil {
-				return err
-			}
-		}
-
-		return nil
-	}
-
-	if r.files == nil {
-		log.Warning("expected more arguments from stdin")
-		return nil
-	}
-
-	fi, err := r.files.NextFile()
+func (req *Request) ParseBodyArgs() error {
+	s, err := req.BodyArgs()
 	if err != nil {
 		return err
 	}
 
-	var any bool
-	scan := bufio.NewScanner(fi)
-	for scan.Scan() {
-		any = true
-		err := f(scan.Text())
-		if err != nil {
-			return err
-		}
-	}
-	if !any {
-		return f("")
+	for s.Scan() {
+		req.Arguments = append(req.Arguments, s.Text())
 	}
 
-	return nil
+	return s.Err()
 }
 
-func getContext(base context.Context, req Request) (context.Context, error) {
-	var (
-		found bool
-		tout  string
-	)
-
-	if optVal := req.Option("timeout"); optVal != nil {
-		var err error
-		tout, found, err = optVal.String()
-		if err != nil {
-			return nil, fmt.Errorf("error parsing timeout option: %s", err)
-		}
-	}
+func (req *Request) SetOption(name string, value interface{}) {
+	optDefs, err := req.Root.GetOptions(req.Path)
+	optDef, found := optDefs[name]
```

Comment:
wait, if GetOptions fails, does it not return nil? This would panic in that case, unless i'm missing something

Reply:
It returns nil, but reading a nil map gives you (<zero>, false): https://play.golang.org/p/vQMkk4nuVN4

Reply:
whoa what. TIL